beamline.rst

```diff
@@ -1,212 +1,379 @@
-Beamline API
-~~~~~~~~~~~~
+General API
+===========
 
-Functionality that incorporates several different instruments or that does
-not necessarily belong to one particular instrument.
+  Functionality (Procedures, Actuators) used generally across the interface.
 
+    # NB I propose using the classes here globally across the interface.
+    #    We need to decide if we want a single namespace for e.g. procedures,
+    #    or we want to have nested dictionaries to divide them by component.
 
 
-**API Functions**
+
+API Functions
+-------------
 
 .. code:: python
 
-   class ProcedureData(NamedTuple):
-    """
-    Describes a "generic" beamline procedure, procedures that are not
-    necassarily present on all beamlines.
-
-    where:
-
-    cmd_id: the string identifying the command, i.e. 'QUICK_REALIGN'
-    running: True if the procedure is running otherwise False
-    args: arguments to pass to the procedure
-    display_name: the name to display in the user interface i.e. "Quick Realign"
-    tool_tip: the tool tip
-    messages: any initial messages to display
-    """
-        cmd_id: str
+   from typing import *
+
+   from enum import Enum
+
+   **# PROCEDURES**
+
+    class ProcedureData(NamedTuple):
+        """
+        Describes a "generic" beamline procedure, procedures that are not
+        necessarily present on all beamlines.
+
+        where:
+
+        name: the string identifying the command, i.e. 'QUICK_REALIGN'
         state: CmdState
-        args: tuple
-        display_name: str
-        tool_tip: str
-        messages: list
+        args: tuple of arguments to pass to the procedure
+        kwargs: dict[str:Any] of keyword arguments with their defaults
+        display_name: the name to display in the user interface i.e. "Quick Realign"
+        tool_tip: the tool tip
+        messages: any initial messages to display
+        """
+
+        name: str   # NB This assumes that only one instance of a command can run at any time
+        state: CmdState
+        args: tuple[str, ...]
+        kwargs: dict[str, Any]
+        display_name: Optional[str] # defaults to name
+        tool_tip: Optional[str]
+        messages: Tuple[str]
+
 
+    def get_procedures() -> OrderedDict[str,ProcedureData]:
+        """
+        :returns: an OrderedDict name:ProcedureData, each describing a procedure,
+                  such as: realignment, anneal
 
-    def get_procedures() -> List[ProcedureData]:
-    """
-    :returns: a list of tuples each describing a beamline procedure,
-              such as: realignment, anneal
+        # NB we may want nested dictionaries if procedures are divided by component
 
-    :rtype: List[ProcedureData]
-    """
+        :rtype: Tuple[ProcedureData]
```

Comment:
Actually the return data could as well be Dict[str, ProcedureData], where the key is the procedure name